swarm/storage/feeds: Renamed MRU to Swarm Feeds

jpeletier opened the pull request:
Renaming MRUs to Swarm Feeds
===

## Abstract
As the Mutable Resource Updates code evolved in the last couple of months, we realized that the end result looked more like a user posting to an update feed than a user "updating a previously created resource".

This PR looks at all Swarm code in excruciating detail, renaming all identifiers and comments so that it reflects this paradigm.

There are no functionality changes in this PR.

## Terms that have been renamed:

* `mru` package is now `feed` and it is located in `/swarm/storage/feed`
* `mru.View` is renamed to `feed.Feed`
* `/bzz-resource:` endpoint is renamed to `/bzz-feed:`
* Swarm command-line `swarm resource` has been renamed to `swarm feed`.

## What objects can I work with from the `feed` package?
You will only need to use three objects:
* `Feed`
* `Query`
* `Request`

### What is a `Feed`?
You can think of a `Feed` as a user's Twitter account, where he/she posts updates about a particular `Topic` (hashtag). In fact, the `Feed` object is simply defined as:
```go
type Feed struct {
	Topic Topic
	User  common.Address
}
```
Users can post about any topic or have topics in common. If you know the user's address and agree on a particular topic, you can then effectively "follow" that user's feed.

### What is a `Query`?
The `Query` object allows you to build a query to browse a particular `Feed`. The default `Query`, obtained with `feed.NewQueryLatest()` will build a `Query` that retrieves the latest update of the given `Feed`.
You can also use `feed.NewQuery()` instead, if you want to build a `Query` to look up an update before a certain date.
Advanced usage of `Query` includes hinting the lookup algorithm for faster lookups. The default hint (`lookup.NoClue`) will have your node track feeds you query frequently and handle hints automatically.

### What is a `Request`?
The `Request` object makes it easy to construct and sign a request to Swarm to update a particular `Feed`. It contains methods to sign and add data.
A `Request` can also be serialized to JSON in case you need your application to delegate signatures, such as having a browser sign a feed update request.

### Example client code:
See for yourself how the new naming improves readability of your code. The following example uses Swarm's client package to retrieve a ready-to-sign `Request`, add data, sign it and post it to Swarm:

```go
// Build a `Feed` object to track a particular user's updates
f := new(feed.Feed)
f.User = signer.Address()
f.Topic, _ = feed.NewTopic("weather",nil)

// Build a `Query` to retrieve a current Request for this feed
query := feeds.NewQueryLatest(&f, lookup.NoClue)

// Retrieve a ready-to-sign request using our query
// (queries can be reused)
request, err := client.GetFeedRequest(query, "")
if err != nil {
    utils.Fatalf("Error retrieving feed status: %s", err.Error())
}

// set the new data
request.SetData([]byte("Weather looks bright and sunny today, we should merge this PR and go out enjoy"))

// sign update
if err = request.Sign(signer); err != nil {
    utils.Fatalf("Error signing feed update: %s", err.Error())
}

// post update
err = client.UpdateFeed(request)
if err != nil {
    utils.Fatalf("Error updating feed: %s", err.Error())
    return
}
```

jpeletier commented on swarm/OWNERS:
```diff
@@ -22,5 +22,5 @@ swarm
 │   ├── encryption ──────── @gbalint, @zelig, @nagydani
 │   ├── mock ────────────── @janos
-│   └── mru ─────────────── @nolash
+│   └── feeds ───────────── @nolash
```
:) what an honor.
Added. Thanks.

zelig commented on swarm/storage/feeds/handler_test.go:
```diff
@@ -266,11 +266,11 @@ func TestSparseUpdates(t *testing.T) {
 	defer os.RemoveAll(datadir)
 
-	// create a new resource
+	// create a new Feed
```
feed

zelig commented on swarm/storage/feeds/handler_test.go:
```diff
@@ -382,7 +382,7 @@ func TestValidator(t *testing.T) {
 
 // tests that the content address validator correctly checks the data
-// tests that resource update chunks are passed through content address validator
+// tests that Feed update chunks are passed through content address validator
```
feed

jpeletier replied:
fixed. thanks

zelig commented on swarm/storage/feeds/handler_test.go:
```diff
@@ -451,20 +451,20 @@ func TestValidatorInStore(t *testing.T) {
 
-// create rpc and resourcehandler
-func setupTest(timeProvider timestampProvider, signer Signer) (rh *TestHandler, datadir string, teardown func(), err error) {
+// create rpc and Feeds Handler
```
feeds handler

zelig commented on swarm/storage/feeds/lookup/lookup.go:
```diff
@@ -15,7 +15,7 @@
 
 /*
-Package lookup defines resource lookup algorithms and provides tools to place updates
+Package lookup defines Feed lookup algorithms and provides tools to place updates
```
feed

zelig commented on swarm/storage/feeds/request.go:
```diff
@@ -24,15 +24,15 @@ import (
 )
 
-// Request represents an update and/or resource create message
+// Request represents a request to sign or signed Feed Update message
```
feed update

zelig commented on swarm/storage/feeds/request_test.go:
```diff
@@ -47,51 +47,49 @@ func areEqualJSON(s1, s2 string) (bool, error) {
 
 // TestEncodingDecodingUpdateRequests ensures that requests are serialized properly
-// while also checking cryptographically that only the owner of a resource can update it.
+// while also checking cryptographically that only the owner of a Feed can update it.
```
feed

zelig commented on swarm/storage/feeds/request_test.go:
```diff
@@ -138,7 +136,7 @@ func TestEncodingDecodingUpdateRequests(t *testing.T) {
 	}
 
-	// Now imagine Bob wants to create an update of his own about the same resource,
+	// Now imagine Bob wants to create an update of his own about the same Feed,
```
feed

zelig commented on swarm/storage/feeds/request_test.go:
```diff
@@ -230,7 +228,7 @@ func TestUpdateChunkSerializationErrorChecking(t *testing.T) {
 	recovered.fromChunk(chunk.Address(), chunk.Data())
 	if !reflect.DeepEqual(recovered, r) {
-		t.Fatal("Expected recovered SignedResource update to equal the original one")
+		t.Fatal("Expected recovered Request update to equal the original one")
```
feed update request

zelig commented on swarm/storage/feeds/request_test.go:
```diff
@@ -250,23 +248,23 @@ func TestReverse(t *testing.T) {
 	signer := newAliceSigner()
 
-	// set up rpc and create resourcehandler
+	// set up rpc and create Feeds handler
```
feeds

zelig commented on swarm/storage/feeds/sign.go:
```diff
@@ -28,7 +28,7 @@ const signatureLength = 65
 type Signature [signatureLength]byte
 
-// Signer signs Mutable Resource update payloads
+// Signer signs Feed update payloads
```
feed

zelig commented on swarm/storage/feeds/sign.go:
```diff
@@ -65,7 +65,7 @@ func (s *GenericSigner) Address() common.Address {
 }
 
-// getUserAddr extracts the address of the resource update signer
+// getUserAddr extracts the address of the Feed update signer
```
feed

zelig commented on cmd/swarm/main.go:
```diff
@@ -346,62 +346,62 @@ func init() {
-					Usage:              "creates and publishes a new Mutable Resource manifest",
-					Description: `creates and publishes a new Mutable Resource manifest pointing to a specified user's updates about a particular topic.
-					The resource topic can be built in the following ways:
+					Usage:              "creates and publishes a new Feed manifest",
```
new feed

zelig commented on cmd/swarm/main.go:
```diff
@@ -346,62 +346,62 @@ func init() {
-					Description: `creates and publishes a new Mutable Resource manifest pointing to a specified user's updates about a particular topic.
-					The resource topic can be built in the following ways:
+					Usage:              "creates and publishes a new Feed manifest",
+					Description: `creates and publishes a new Feed manifest pointing to a specified user's updates about a particular topic.
```
new feed

zelig commented on cmd/swarm/main.go:
```diff
@@ -346,62 +346,62 @@ func init() {
-					Usage:              "obtains information about an existing Mutable Resource",
-					Description: `obtains information about an existing Mutable Resource
+					Usage:              "obtains information about an existing Swarm Feed",
+					Description: `obtains information about an existing Swarm Feed
```
Swarm feed

zelig commented on swarm/api/api.go:
```diff
@@ -956,94 +957,93 @@ func (a *API) BuildDirectoryTree(ctx context.Context, mhash string, nameresolver
-// ResourceLookup finds mutable resource updates at specific periods and versions
-func (a *API) ResourceLookup(ctx context.Context, query *mru.Query) ([]byte, error) {
-	_, err := a.resource.Lookup(ctx, query)
+// FeedsLookup finds Swarm Feeds Updates at specific points in time, or the latest update
```
Swarm feed updates

zelig commented on swarm/api/api.go:
```diff
@@ -956,94 +957,93 @@ func (a *API) BuildDirectoryTree(ctx context.Context, mhash string, nameresolver
-// ResourceNewRequest creates a Request object to update a specific mutable resource
-func (a *API) ResourceNewRequest(ctx context.Context, view *mru.View) (*mru.Request, error) {
-	return a.resource.NewRequest(ctx, view)
+// FeedsNewRequest creates a Request object to update a specific Feed
```
specific feed

zelig commented on swarm/api/api.go:
```diff
@@ -956,94 +957,93 @@ func (a *API) BuildDirectoryTree(ctx context.Context, mhash string, nameresolver
-// ResourceHashSize returned the size of the digest produced by the Mutable Resource hashing function
-func (a *API) ResourceHashSize() int {
-	return a.resource.HashSize
+// FeedsHashSize returned the size of the digest produced by Swarm Feeds' hashing function
```
feeds'

zelig commented on swarm/api/client/client.go:
```diff
@@ -701,18 +700,18 @@ func (c *Client) getResource(query *mru.Query, manifestAddressOrDomain string, m
 
-// GetResourceMetadata returns a structure that describes the Mutable Resource
-// manifestAddressOrDomain is the address you obtained in CreateResource or an ENS domain whose Resolver
+// GetFeedRequest returns a structure that describes the referenced Feed status
```
feed status

zelig commented on swarm/api/http/server_test.go:
```diff
@@ -206,8 +165,8 @@ func TestBzzResourceMultihash(t *testing.T) {
 
-// Test resource updates using the raw update methods
-func TestBzzResource(t *testing.T) {
+// Test Swarm Feeds using the raw update methods
```
feeds

zelig commented on swarm/network/README.md:
```diff
@@ -57,7 +57,7 @@ receipts for a deleted chunk easily to refute their challenge.
 keep track of syncing state across sessions, historical syncing state should survive restart
 - extra data structures to support syncing should be kept at minimum
-- syncing is organized separately for chunk types (resource update v content chunk)
+- syncing is organized separately for chunk types (Swarm Feed Updates v regular content chunk)
```
this is not true. it is handled not separately

jpeletier replied:
changed to: *syncing is **not** organized separately for chunk types (Swarm feed updates v regular content chunk)*

However, take into account I was just changing terms here, not really changing the meaning. `swarm/network` owner should be aware of this.

zelig commented on swarm/pss/notify/notify.go:
```diff
@@ -317,7 +317,7 @@ func (c *Controller) handleStartMsg(msg *Msg, keyid string) (err error) {
 	}
 
-	// TODO this is set to zero-length byte pending decision on protocol for initial message, whether it should include message or not, and how to trigger the initial message so that current state of MRU is sent upon subscription
+	// TODO this is set to zero-length byte pending decision on protocol for initial message, whether it should include message or not, and how to trigger the initial message so that current state of Swarm Feed is sent upon subscription
```
feed

zelig commented on swarm/storage/feeds/error.go:
```diff
@@ -35,7 +35,7 @@ const (
 )
 
-// Error is a the typed error object used for Mutable Resources
+// Error is a the typed error object used for Swarm Feeds
```
feeds

zelig commented on swarm/storage/feeds/error.go:
```diff
@@ -47,12 +47,12 @@ func (e *Error) Error() string {
 }
 
-// NewError creates a new Mutable Resource Error object with the specified code and custom error message
+// NewError creates a new Swarm Feeds Error object with the specified code and custom error message
```
feeds

zelig commented on swarm/storage/feeds/feed.go:
```diff
@@ -25,21 +25,21 @@ import (
 
-// View represents a particular user's view of a resource
-type View struct {
+// Feed represents a particular user's stream of updates on a Topic
```
topic

zelig commented on swarm/storage/feeds/feed.go:
```diff
@@ -48,55 +48,55 @@ func (u *View) mapKey() uint64 {
-func (u *View) binaryPut(serializedData []byte) error {
-	if len(serializedData) != viewLength {
-		return NewErrorf(ErrInvalidValue, "Incorrect slice size to serialize View. Expected %d, got %d", viewLength, len(serializedData))
+// binaryPut serializes this Feed instance into the provided slice
```
feed

zelig commented on swarm/storage/feeds/feed.go:
```diff
@@ -48,55 +48,55 @@ func (u *View) mapKey() uint64 {
+// binaryPut serializes this Feed instance into the provided slice
+func (f *Feed) binaryPut(serializedData []byte) error {
+	if len(serializedData) != feedLength {
+		return NewErrorf(ErrInvalidValue, "Incorrect slice size to serialize Feed. Expected %d, got %d", feedLength, len(serializedData))
```
feed

zelig commented on swarm/storage/feeds/feed.go:
```diff
@@ -48,55 +48,55 @@ func (u *View) mapKey() uint64 {
-		return NewErrorf(ErrInvalidValue, "Incorrect slice size to read View. Expected %d, got %d", viewLength, len(serializedData))
+func (f *Feed) binaryGet(serializedData []byte) error {
+	if len(serializedData) != feedLength {
+		return NewErrorf(ErrInvalidValue, "Incorrect slice size to read Feed. Expected %d, got %d", feedLength, len(serializedData))
```
feed

zelig commented on swarm/storage/feeds/handler.go:
```diff
@@ -14,9 +14,9 @@
-// Handler is the API for Mutable Resources
-// It enables creating, updating, syncing and retrieving resources and their update data
-package mru
+// Handler is the API for Feeds
```
feeds

zelig commented on swarm/storage/feeds/handler.go:
```diff
@@ -52,35 +52,35 @@ var hashPool sync.Pool
 }
 
-// SetStore sets the store backend for the Mutable Resource API
+// SetStore sets the store backend for the Swarm Feeds API
```
feeds

zelig commented on swarm/storage/feeds/handler.go:
```diff
@@ -89,81 +89,78 @@ func (h *Handler) Validate(chunkAddr storage.Address, data []byte) bool {
-func (h *Handler) GetContent(view *View) (storage.Address, []byte, error) {
-	if view == nil {
-		return nil, nil, NewError(ErrInvalidValue, "view is nil")
+// GetContent retrieves the data payload of the last synced update of the Feed
```
of the feed

zelig commented on swarm/storage/feeds/handler.go:
```diff
@@ -267,32 +264,32 @@ func (h *Handler) Update(ctx context.Context, r *Request) (updateAddr storage.Ad
-	h.resourceLock.Lock()
-	defer h.resourceLock.Unlock()
-	h.resources[mapKey] = rsrc
+// Sets the feed update cache value for the given Feed
```
given feed

zelig commented on swarm/storage/feeds/handler_test.go:
```diff
@@ -83,18 +83,18 @@ func TestResourceHandler(t *testing.T) {
 	defer teardownTest()
 
-	// create a new resource
+	// create a new Feed
```
feed

zelig commented on swarm/storage/feeds/handler_test.go:
```diff
@@ -83,18 +83,18 @@ func TestResourceHandler(t *testing.T) {
 
-	topic, _ := NewTopic("Mess with mru code and see what ghost catches you", nil)
-	view := View{
+	topic, _ := NewTopic("Mess with Swarm Feeds code and see what ghost catches you", nil)
```
feeds

jpeletier commented on swarm/api/http/server.go:
```diff
@@ -458,66 +457,35 @@ func (s *Server) HandleDelete(w http.ResponseWriter, r *http.Request) {
 	fmt.Fprint(w, newKey)
 }
 
-// Parses a resource update post url to corresponding action
```
old orphan code not used by any code any more. It has been sitting in there unused for months now.

jpeletier commented on swarm/api/http/server.go:
```diff
@@ -527,30 +495,30 @@ func (s *Server) HandlePostResource(w http.ResponseWriter, r *http.Request) {
 		}
 		// the key to the manifest will be passed back to the client
-		// the client can access the view  directly through its resourceView member
+		// the client can access the Feed  directly through its Feed member
```
fixed. Thanks.

jpeletier commented on .github/CODEOWNERS:
```diff
@@ -27,6 +27,6 @@ swarm/services                  @zelig
 swarm/storage/encryption        @gbalint @zelig @nagydani
 swarm/storage/mock              @janos
-swarm/storage/mru               @nolash
+swarm/storage/feeds             @nolash
```
done. Thanks.

jpeletier commented on swarm/network/README.md:
```diff
@@ -37,7 +37,7 @@ Using the streamer logic, various stream types are easy to implement:
   * historical syncing
 * simple retrieve requests and deliveries
-* mutable resource updates streams
+* Swarm Feeds streams
```
fixed. thanks

jpeletier commented on swarm/storage/feeds/binaryserializer.go:
```diff
@@ -14,7 +14,7 @@
 // along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
 
-package mru
+package feeds
```
"Swarm Feeds" sounded better to me from a "marketing" standpoint, i.e., *I am using Swarm Feeds in my project so my users can collaborate"*

Also, when instantiating the `Handler`, plural seemed better: `fh:= new(feeds.Handler)` (the handler of all the feeds). Singular seemed like the handler of 1 feed.

I'll rename it, the "marketing name" in documentation does not have to be exactly 1:1 with the code name.

jpeletier commented:
@zelig 
* Renamed package to `feed`
* Renamed all conflicting vars to `f` or `fd`
* Fixed all capitalized `Feed` + regex searched and found more, also fixed.
* Rebased.

zelig commented on .github/CODEOWNERS:
```diff
@@ -27,6 +27,6 @@ swarm/services                  @zelig
 swarm/storage/encryption        @gbalint @zelig @nagydani
 swarm/storage/mock              @janos
-swarm/storage/mru               @nolash
+swarm/storage/feeds             @nolash @jpeletier
```
feed
:)

jpeletier replied:
:+1:

zelig commented on swarm/OWNERS:
```diff
@@ -22,5 +22,5 @@ swarm
 │   ├── encryption ──────── @gbalint, @zelig, @nagydani
 │   ├── mock ────────────── @janos
-│   └── mru ─────────────── @nolash
+│   └── feeds ───────────── @nolash, @jpeletier
```
feed

zelig commented on swarm/swarm.go:
```diff
@@ -50,7 +50,7 @@ import (
 	"github.com/ethereum/go-ethereum/swarm/storage"
 	"github.com/ethereum/go-ethereum/swarm/storage/mock"
-	"github.com/ethereum/go-ethereum/swarm/storage/mru"
+	"github.com/ethereum/go-ethereum/swarm/storage/feed"
```
formatting! 
https://travis-ci.org/ethereum/go-ethereum/jobs/436008109#L511

jpeletier replied:
fixed. Sorry.
